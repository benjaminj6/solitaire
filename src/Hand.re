open Types;
let component = ReasonReact.statelessComponent("Hand");

module Styles = {
  open Css;

  let wrapper = style([
    display(`flex),
    justifyContent(`flexEnd),
  ]);

  let card = i => {
    let overlap = i == 0 ? `zero : rem(-2.0);

    [marginLeft(overlap), maxWidth(pct(50.0))];
  };
};

let make = (~hand, ~onClickCard, _children) => {
  ...component,
  render: _self => {
    let displayedCards =
      switch (hand) {
      | [] => []
      | [a] => [a]
      | [a, b] => [b, a]
      | [a, b, c, ..._rest] => [c, b, a]
      };

    /*
       TODO -- may need to create an internal grid with the layers here...
       It would be 50 / 50 inheriting the grid-gap of its parent.
       this could mean it's more worth it to split the CardStack component just
       up into the Hand & the Foundation Stacks
     */
    <div className=Styles.wrapper>
      {
        displayedCards
        |> List.mapi((i, card) =>
             <Card
               key={card.id |> string_of_int}
               card
               styles={Styles.card(i)}
               onClick={(~card) => onClickCard(~location=Hand, ~card=Some(card))}
             />
           )
        |> Array.of_list
        |> ReasonReact.array
      }
    </div>;
  },
};