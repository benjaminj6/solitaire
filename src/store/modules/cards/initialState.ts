import * as range from 'lodash/fp/range';
import { Card, CardRank, CardRankNumber, Deck, State, Suit } from './types';

type GetRankFromRemainder = (x: CardRankNumber) => CardRank;
export const getRankFromRemainder: GetRankFromRemainder = num => {
  switch (num) {
    case '0':
      return 'K';
    case '1':
      return 'Q';
    case '2':
      return 'J';
    case '12':
      return 'A';
    default:
      return `${13 - +num}` as CardRank;
  }
};

type GetInitialSuit = (x: Suit) => Deck;
export const getInitialSuit: GetInitialSuit = suitName => {
  return range(1, 14)
    .map((num: number) => {
      const value = `${num % 13}` as CardRankNumber;
      const rank = getRankFromRemainder(value);

      return {
        id: `${suitName}-${rank}`,
        rank,
        suit: suitName,
        value: 13 - +value,
        visible: false,
      };
    })
    .sort((first: Card, second: Card) => first.value - second.value)
    .reduce((obj: Deck, card: Card) => {
      obj[card.id] = card;
      return obj;
    }, {});
};

type GetInitialDeck = () => Deck;
export const getInitialDeck: GetInitialDeck = () => {
  const suits = ['hearts', 'spades', 'diamonds', 'clubs'];
  return suits.reduce(
    (obj: Deck, suitName: Suit) => ({
      ...obj,
      ...getInitialSuit(suitName),
    }),
    {}
  );
};

export default {
  deck: getInitialDeck(),
  shuffledDeck: [],
} as State;
