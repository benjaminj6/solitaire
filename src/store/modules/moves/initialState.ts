import { State } from './types';

const initialState: State = {
  nextMove: {},
  moves: {},
  history: [],
};

export default initialState;