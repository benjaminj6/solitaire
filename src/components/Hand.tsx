import * as React from 'react';
import Card from './Card';
import { Hand, Deck } from '../store/modules/cards/types';
import * as cx from 'classnames';
import { last, negate as not } from 'lodash';

type Props = {
  className?: string;
  style?: { [key: string]: string };
  hand: Hand;
  deck: Deck;
};

const isLast = (i: number, arr: any[]) => i + 1 === arr.length;
const isNotLast = not(isLast);

const HandComponent: React.SFC<Props> = ({ hand, deck, className, style }) => (
  <div
    className={cx(className, 'grid', 'grid-template-columns', 'h-24')}
    style={{ ...style, '--grid-template-columns': '1fr 1fr' }}
  >
    {hand.length
      ? hand.slice(-3).map((cardId, i, arr) => (
          <Card
            disabled={isNotLast(i, arr)}
            className={cx(
              'grid-column',
              'grid-row',
              'relative',
              {
                'right-neg-1/4': i !== 0 && isNotLast(i, arr),
              },
              `z-${i + 1}0` // z-10, z-20, z-30, etc.
            )}
            style={{
              '--grid-column': isLast(i, arr) ? '2' : '1',
              '--grid-row': '1',
            }}
            key={cardId}
            {...deck[cardId]}
          />
        ))
      : null}
  </div>
);

export default HandComponent;
