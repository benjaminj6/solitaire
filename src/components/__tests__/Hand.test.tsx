import * as React from 'react';
import { render } from 'react-testing-library';
import { createRenderer } from 'react-test-renderer/shallow';
import Hand from '../Hand';
import initialState from '../../store/modules/cards/initialState';

const renderer = createRenderer();

describe('Hand', () => {
  test('should render', () => {
    const tree = renderer.render(<Hand
      hand={['spades-2', 'diamonds-4', 'hearts-9', 'clubs-K']}
      deck={initialState.deck}
    />);

    expect(tree).toMatchSnapshot();
  });

  test('should render nothing if there is no length to the hand', () => {
    const tree = renderer.render(<Hand deck={initialState.deck} hand={[]} />);

    expect(tree).toMatchSnapshot();
  });
});
