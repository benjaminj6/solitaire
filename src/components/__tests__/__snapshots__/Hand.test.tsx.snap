// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hand should render 1`] = `
<div
  className="grid grid-template-columns h-24"
  style={
    Object {
      "--grid-template-columns": "1fr 1fr",
    }
  }
>
  <Card
    className="grid-column grid-row relative z-10"
    disabled={true}
    id="diamonds-4"
    rank="4"
    style={
      Object {
        "--grid-column": "1",
        "--grid-row": "1",
      }
    }
    suit="diamonds"
    value={4}
    visible={false}
  />
  <Card
    className="grid-column grid-row relative right-neg-1/4 z-20"
    disabled={true}
    id="hearts-9"
    rank="9"
    style={
      Object {
        "--grid-column": "1",
        "--grid-row": "1",
      }
    }
    suit="hearts"
    value={9}
    visible={false}
  />
  <Card
    className="grid-column grid-row relative z-30"
    disabled={false}
    id="clubs-K"
    rank="K"
    style={
      Object {
        "--grid-column": "2",
        "--grid-row": "1",
      }
    }
    suit="clubs"
    value={13}
    visible={false}
  />
</div>
`;

exports[`Hand should render nothing if there is no length to the hand 1`] = `
<div
  className="grid grid-template-columns h-24"
  style={
    Object {
      "--grid-template-columns": "1fr 1fr",
    }
  }
/>
`;
